File: packages/platform/src/ipc/publish-batch.ts
/**
 * 批量发布 IPC 封装 — 提交任务 + 查询进度。
 *
 * 后端：壳层 `commands/publish_batch.rs`（InMemoryBatchStore + app::publish::BatchService）。
 *
 * @author agent
 * @created 2026-08-13
 */

import { call } from "./invoke";

/** 单账号发布统计（与 Rust `BatchAccountStatus` 对齐）。 */
export interface BatchAccountStatus {
  account_id: string;
  total: number;
  success: number;
  failed: number;
  publishing: number;
  pending: number;
  sync_status: string;
  sync_message: string;
  sync_total_count: number;
  sync_saved_count: number;
}

/** 批量发布任务（与 Rust `BatchTask` 对齐）。 */
export interface BatchTask {
  batch_id: string;
  owner_id: number;
  account_ids: string[];
  material_ids: number[];
  total: number;
  success: number;
  failed: number;
  publishing: number;
  pending: number;
  finished: boolean;
  account_statuses: BatchAccountStatus[];
}

const OWNER_ID = 1; // 桌面单用户；多用户时由登录态注入

/** 提交批量发布任务（后台执行，返回任务快照）。 */
export function publishBatchSubmit(
  accountIds: string[],
  materialIds: number[],
): Promise<BatchTask> {
  return call<BatchTask>("publish_batch_submit", {
    request: { owner_id: OWNER_ID, account_ids: accountIds, material_ids: materialIds },
  });
}

/** 查询批量发布进度。 */
export function publishBatchStatus(batchId: string): Promise<BatchTask | null> {
  return call<BatchTask | null>("publish_batch_status", {
    request: { owner_id: OWNER_ID, batch_id: batchId },
  });
}

File: apps/desktop/src-tauri/src/platforms/xianyu/ipc/publish_batch.rs
//! 批量发布 Tauri commands。

use crate::platforms::xianyu::adapter::InMemoryPublishGateway;
use crate::platforms::xianyu::persist::InMemoryBatchStore;
use crate::shared::ipc::IpcResponse;
use app::publish::{BatchService, BatchStore, BatchTask, PublishRequest, PublishService};
use common;
use serde::Deserialize;
use std::sync::Arc;
use tauri::State;

pub struct BatchPublishHandle {
    pub store: Arc<InMemoryBatchStore>,
    pub gateway: Arc<InMemoryPublishGateway>,
}

#[derive(Debug, Deserialize)]
pub struct BatchSubmitRequest {
    pub owner_id: i64,
    pub account_ids: Vec<String>,
    pub material_ids: Vec<i64>,
}

#[derive(Debug, Deserialize)]
pub struct BatchStatusRequest {
    pub owner_id: i64,
    pub batch_id: String,
}

#[tauri::command]
pub async fn publish_batch_submit(
    state: State<'_, BatchPublishHandle>,
    request: BatchSubmitRequest,
) -> common::DingDaResult<IpcResponse<BatchTask>> {
    let batch_id = format!("batch-{}", uuid_fragment());
    let service = BatchService::new(state.store.as_ref());
    let task = service.submit(
        &batch_id,
        request.owner_id,
        &request.account_ids,
        &request.material_ids,
    )?;

    let store = state.store.clone();
    let gateway = state.gateway.clone();
    let task_snapshot = task.clone();
    let account_ids = task.account_ids.clone();
    let material_ids = task.material_ids.clone();
    tauri::async_runtime::spawn(async move {
        let publish = PublishService::new(gateway.as_ref());
        let mut task = task_snapshot;
        for account_id in &account_ids {
            for material_id in &material_ids {
                let item = serde_json::json!({
                    "title": format!("批量发布素材 #{material_id}"),
                    "description": "",
                    "price": 0,
                    "images": []
                });
                let result = publish
                    .execute(&PublishRequest {
                        user_id: task.owner_id,
                        account_id: account_id.clone(),
                        item,
                        material_id: Some(*material_id),
                    })
                    .await;
                task.record(account_id, result.success);
                task.mark_sync(
                    account_id,
                    "skipped",
                    "内存网关：批量发布商品同步由 sidecar 执行",
                    0,
                    0,
                );
                let _ = store.update_task(&task);
            }
        }
        if !task.finished {
            task.finished = true;
        }
        let _ = store.update_task(&task);
    });

    Ok(IpcResponse::ok(task))
}

#[tauri::command]
pub fn publish_batch_status(
    state: State<'_, BatchPublishHandle>,
    request: BatchStatusRequest,
) -> common::DingDaResult<IpcResponse<Option<BatchTask>>> {
    let service = BatchService::new(state.store.as_ref());
    let result = service.status(request.owner_id, &request.batch_id)?;
    Ok(IpcResponse::ok(result))
}

fn uuid_fragment() -> String {
    use std::time::{SystemTime, UNIX_EPOCH};
    let millis = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|duration| duration.as_millis())
        .unwrap_or(0);
    format!("{:x}", millis % 1_000_000_000)
}
